test/e2e/node/security_context.go

```diff
@@ -146,37 +147,59 @@ var _ = SIGDescribe("Security Context", func() {
 		pod.Spec.Containers[0].Command = []string{"grep", "ecc", "/proc/self/status"}
 		f.TestContainerOutput(v1.SeccompPodAnnotationKey, pod, 0, []string{"0"}) // seccomp disabled
 	})
 
-	ginkgo.It("should support seccomp alpha unconfined annotation on the pod [Feature:Seccomp] [LinuxOnly]", func() {
-		// TODO: port to SecurityContext as soon as seccomp is out of alpha
+	ginkgo.It("should support seccomp unconfined on the pod [LinuxOnly]", func() {
 		pod := scTestPod(false, false)
-		pod.Annotations[v1.SeccompPodAnnotationKey] = "unconfined"
+		pod.Spec.SecurityContext.SeccompProfile = &v1.SeccompProfile{Type: v1.SeccompProfileUnconfined}
 		pod.Spec.Containers[0].Command = []string{"grep", "ecc", "/proc/self/status"}
 		f.TestContainerOutput(v1.SeccompPodAnnotationKey, pod, 0, []string{"0"}) // seccomp disabled
 	})
 
-	ginkgo.It("should support seccomp alpha runtime/default annotation [Feature:Seccomp] [LinuxOnly]", func() {
-		// TODO: port to SecurityContext as soon as seccomp is out of alpha
+	ginkgo.It("should support seccomp runtime/default on the container [LinuxOnly]", func() {
 		pod := scTestPod(false, false)
-		pod.Annotations[v1.SeccompContainerAnnotationKeyPrefix+"test-container"] = v1.SeccompProfileRuntimeDefault
+		pod.Spec.SecurityContext.SeccompProfile = &v1.SeccompProfile{Type: v1.SeccompProfileUnconfined}
+		pod.Spec.Containers[0].SecurityContext.SeccompProfile = &v1.SeccompProfile{Type: v1.SeccompProfileTypeDefault}
 		pod.Spec.Containers[0].Command = []string{"grep", "ecc", "/proc/self/status"}
 		f.TestContainerOutput(v1.SeccompPodAnnotationKey, pod, 0, []string{"2"}) // seccomp filtered
 	})
 
-	ginkgo.It("should support seccomp default which is unconfined [Feature:Seccomp] [LinuxOnly]", func() {
-		// TODO: port to SecurityContext as soon as seccomp is out of alpha
+	ginkgo.It("should support seccomp runtime/default on the pod [LinuxOnly]", func() {
+		pod := scTestPod(false, false)
+		pod.Spec.SecurityContext.SeccompProfile = &v1.SeccompProfile{Type: v1.SeccompProfileTypeDefault}
+		pod.Spec.Containers[0].Command = []string{"grep", "ecc", "/proc/self/status"}
+		f.TestContainerOutput(v1.SeccompPodAnnotationKey, pod, 0, []string{"2"}) // seccomp filtered
+	})
+
+	ginkgo.It("should support seccomp default which is unconfined [LinuxOnly]", func() {
 		pod := scTestPod(false, false)
 		pod.Spec.Containers[0].Command = []string{"grep", "ecc", "/proc/self/status"}
 		f.TestContainerOutput(v1.SeccompPodAnnotationKey, pod, 0, []string{"0"}) // seccomp disabled
 	})
+
+	ginkgo.It("should not support seccomp localhost profile on the pod if empty [LinuxOnly]", func() {
+		pod := scTestPod(false, false)
+		pod.Spec.SecurityContext.SeccompProfile = &v1.SeccompProfile{Type: v1.SeccompProfileLocalhost}
+
+		_, err := f.PodClientNS(f.Namespace.Name).PodInterface.Create(context.TODO(), pod, metav1.CreateOptions{})
+		framework.ExpectError(err)
+	})
+
```

Comment:
@saschagrunert What's your thoughts about checking whether the pod level seccomp applies to the sandbox container?

Reply:
Hm, how would we verify that if not from a running container? (like https://github.com/kubernetes/kubernetes/pull/91442/commits/db5843d1b3a23299e3f5fa5b4d121db57d479a3c#diff-d396c518f484bf943b4f92eec90c708bR175)

I think we could make the tests more valuable when bringing real localhost profiles into the game. We could change the test bootstrap process to place the profiles in the nodes. But now I'm wondering which test we would change for that. Probably `pull-kubernetes-e2e-gce` and we could scope the tests with something like `[Seccomp Localhost]`.

Reply:
> Hm, how would we verify that if not from a running container? (like [db5843d#diff-d396c518f484bf943b4f92eec90c708bR175](https://github.com/kubernetes/kubernetes/commit/db5843d1b3a23299e3f5fa5b4d121db57d479a3c#diff-d396c518f484bf943b4f92eec90c708bR175))

To test we could use a localhost profile at pod level that would block the sandbox requirements - or simply a default action of ERRNO. Last time I mapped sandbox's profile required:
https://github.com/pjbgf/kubernetes/blob/a3bf173f20904a8ee70bbc9526dd5b605c511757/cmd/kubelet/app/options/container_runtime.go#L46

Then ensure that the pod would run if the sandbox was not blocked, by overwritting the seccomp of the other container running in that pod by setting its SeccompProfile to Unconfined or RuntimeDefault. If pod fails to run, the seccomp was applied to the sandbox.

> I think we could make the tests more valuable when bringing real localhost profiles into the game. We could change the test bootstrap process to place the profiles in the nodes. But now I'm wondering which test we would change for that. Probably `pull-kubernetes-e2e-gce` and we could scope the tests with something like `[Seccomp Localhost]`.

Sounds good.

Reply:
I think we have the "overwrite pod sandbox profile in container" scenario already covered, but with localhost profiles it would be better to test. I'll check out what is needed to add sample profile to the nodes...

Reply:
Just to clarify the test I mentioned:

```
apiVersion: v1
kind: Pod
metadata:
  name: pod-level-container-perms
  annotations:
    seccomp.security.alpha.kubernetes.io/pod: 'localhost/blockall.json'
    container.seccomp.security.alpha.kubernetes.io/matey-container: 'unconfined'
spec:
  containers:
  - name: matey-container
    image: alpine
    command: ["sh"]
    args: ["-c", "echo hi && sleep 120"]
```
Using profile:
```
{
    "defaultAction": "SCMP_ACT_ERRNO"
}
```

This fails to start - which is the expected behaviour - with the following error:

```
4s          Warning   FailedCreatePodSandBox    pod/pod-level-container-perms   Failed to create pod sandbox: rpc error: code = Unknown desc = failed to start sandbox container for pod "pod-level-container-perms": Error response from daemon: OCI runtime start failed: cannot start a container that has stopped: unknown
```

If the seccomp is not applied to the sandbox container, the pod would execute normally.